src/modm/math/filter/moving_average.hpp

```diff
@@ -85,16 +83,21 @@ namespace modm
 
 // ----------------------------------------------------------------------------
 template<typename T, std::size_t N>
-modm::filter::MovingAverage<T, N>::MovingAverage(const T& initialValue) :
-	index(0), sum(N * initialValue)
+modm::filter::MovingAverage<T, N>::MovingAverage(const T& initialValue)
 {
-	for (Index i = 0; i < N; ++i) {
-		buffer[i] = initialValue;
-	}
+	reset(initialValue);
+}
+
+// ----------------------------------------------------------------------------
+template<typename T, std::size_t N>
+void
+modm::filter::MovingAverage<T, N>::reset(const T& input)
+{
+	std::fill(std::begin(buffer), std::end(buffer), input);
+	sum = N * input;
 }
```

Comment:
Only if you're motivated: Do you think this function and the constructor can be marked as `constexpr`?
Then the sum and buffer may be placed in pure `.data` and copied during startup for static allocation, rather than being constructed at runtime (still required for dynamic allocation of course).

Reply:
I also think it's faster if you copy the argument for this ie. just `T` not `const T&`. On Cortex-M this forces the argument to be pushed onto the stack and then a pointer is passed due to calling convention, which is slower even for a 64-bit double.

Reply:
The **p**ass **b**y **c**opy **pbc** VS  **p**ass **b**y **r**eference **pbr** topic still confuses me

- In templates T may be small or big. thus calls for **pbc** or **pbr**
- STL uses **pbr** all around. I've once benchmarked on ATmega: std::min(T&, T&) VS modm::min(T&, T&) VS modm::min(T, T) and found the STL-**pbr** variant being the fastest with little exceptions. So i concluded to myself, the compiler is smart enough to optimize **pbr**declarations in beneficial cases.

Wish i would be better trained in tools like _compiler explorer_ to control, if and how things optimize :/

Reply:
In the case of `MovingAverage` 99% T falls in the category `pbc` so let's correct this.